cloudsync_gdrive.py

```diff
@@ -312,56 +326,103 @@ def events(self) -> Generator[Event, None, None]:  # pylint: disable=too-many-lo
             new_cursor = response.get('newStartPageToken', None)
             for change in response.get('changes'):
                 log.debug("got event %s", change)
+                event = self._convert_to_event(change, new_cursor)
+                log.debug("converted event %s as %s", change, event)
+                filter_result = self._filter_event(event)
+                if filter_result == EventFilter.IGNORE:
+                    if event:
+                        log.debug("ignore event: %s %s %s", event.path, event.oid, event.exists)
+                    continue
+                if filter_result == EventFilter.WALK:
+                    log.debug("directory created in or renamed into root - walking: %s", event.path)
+                    try:
+                        yield from self.walk_oid(event.oid)
+                    except CloudFileNotFoundError:
+                        pass
+                yield event
 
-                # {'kind': 'drive#change', 'type': 'file', 'changeType': 'file', 'time': '2019-07-23T16:57:06.779Z',
-                # 'removed': False, 'fileId': '1NCi2j1SjsPUTQTtaD2dFNsrt49J8TPDd', 'file': {'kind': 'drive#file',
-                # 'id': '1NCi2j1SjsPUTQTtaD2dFNsrt49J8TPDd', 'name': 'dest', 'mimeType': 'application/octet-stream'}}
-
-                # {'kind': 'drive#change', 'type': 'file', 'changeType': 'file', 'time': '2019-07-23T20:02:14.156Z',
-                # 'removed': True, 'fileId': '1lhRe0nDplA6I5JS18642rg0KIbYN66lR'}
-
-                ts = arrow.get(change.get('time')).float_timestamp
-                oid = change.get('fileId')
-                exists = None
-                if 'removed' in change.keys():
-                    exists = not change.get('removed')
-
-                fil = change.get('file')
-                if fil:
-                    if fil.get('mimeType') == self._folder_mime_type:
-                        otype = DIRECTORY
-                    else:
-                        otype = FILE
-                else:
-                    otype = NOTKNOWN
-
-                ohash = None
-                path = self._path_oid(oid, use_cache=False)
+            if new_cursor and page_token and new_cursor != page_token:
+                self.__cursor = new_cursor
+            page_token = response.get('nextPageToken')
 
-                event = Event(otype, oid, path, ohash, exists, ts, new_cursor=new_cursor)
+    def _convert_to_event(self, change, new_cursor):
+        ts = arrow.get(change.get('time')).float_timestamp
+        oid = change.get('fileId')
+        exists = None
```

Comment:
what does event.exists = None mean?

Reply:
This means we don't know if it exists or doesn't exist and that should be determined by `get_latest`. This needs to be supported and is related to the cloudsync PR Erik put up yesterday https://github.com/AtakamaLLC/cloudsync/pull/329/files. For gdrive specifically, we can avoid the an unnecessary API hit for each event if we leave exists as None and wait for get_latest to populate this field